lib/tcpip/tcpip_error_handler.h

```diff
@@ -0,0 +1,367 @@
+	__CriticalSectionFlagLock.lock.lockWord = 0;
+	__CriticalSectionFlagLock.depth         = 0;
+	__SuspendFlagLock.lock.lockWord         = 0;
+	__SuspendFlagLock.depth                 = 0;
+	memset(&sealedSocketsListLock, 0, sizeof(FlagLockPriorityInherited));
+
+	// Restart the network stack. This resets the startup state before
+	// calling `network_start`.
+	DebugErrorHandler::log("Restarting the network stack.");
+	restartState |= IpThreadKicked;
+	network_restart();
+
+	// We do not reset `restartState` here, the network thread will take
+	// care of it when the TCP/IP stack is done reseting.
+}
+
+extern void ip_thread_entry(void);
+
+extern "C" ErrorRecoveryBehaviour
+compartment_error_handler(ErrorState *frame, size_t mcause, size_t mtval)
+{
+	auto threadID = thread_id_get();
+	if (mcause == priv::MCAUSE_CHERI)
+	{
+		auto [exceptionCode, registerNumber] =
+		  CHERI::extract_cheri_mtval(mtval);
+		// The thread entry point is called with a NULL return address so the
+		// cret at the end of the entry point function will trap if it is
+		// reached. We don't want to treat this as an error but thankfully we
+		// detect it quite specifically by checking for all of:
+		// 1) CHERI cause is tag violation
+		// 2) faulting register is CRA
+		// 3) value of CRA is NULL
+		// 4) we've reached the top of the thread's stack
+		Capability stackCapability{
+		  frame->get_register_value<CHERI::RegisterNumber::CSP>()};
+		Capability returnCapability{
+		  frame->get_register_value<CHERI::RegisterNumber::CRA>()};
+		if (registerNumber == CHERI::RegisterNumber::CRA &&
+		    returnCapability.address() == 0 &&
+		    exceptionCode == CHERI::CauseCode::TagViolation &&
+		    stackCapability.top() == stackCapability.address())
+		{
+			// looks like thread exit -- just log it then ForceUnwind
+			DebugErrorHandler::log(
+			  "Thread exit CSP={}, PCC={}", stackCapability, frame->pcc);
+		}
+		else if (exceptionCode == CHERI::CauseCode::None)
+		{
+			// An unwind occurred from a called compartment, just resume.
+			return ErrorRecoveryBehaviour::InstallContext;
+		}
+		else
+		{
+			// An unexpected error -- log it and restart the stack.
+			// Note: handle CZR differently as `get_register_value`
+			// will return a nullptr which we cannot dereference.
+			DebugErrorHandler::log(
+			  "{} error at {} (return address: {}), with capability register "
+			  "{}: {} by thread {}",
+			  exceptionCode,
+			  frame->pcc,
+			  frame->get_register_value<CHERI::RegisterNumber::CRA>(),
+			  registerNumber,
+			  registerNumber == CHERI::RegisterNumber::CZR
+			    ? nullptr
+			    : *frame->get_register_value(registerNumber),
+			  threadID);
+
+			// TODO before running the reset function we should go
+			// to the top of the stack to ensure that we do not run
+			// out of stack space while executing the error
+			// handler.
+
+			// Reset the network stack state.
+			reset_network_stack_state();
```

Comment:
I think the correct thing to do is to make the IP thread call this function whenever it's doing a restart and have the code below set the PCC to a pointer to the run loop function with ca0 set to 1 telling it to do the reset.  Then the cleanup code always runs on a known thread.

Reply:
As we discussed, it would indeed be nice to move as much as we can of the reset to the context of the IP thread, however some bits have to stay here to be able to unblock the IP thread in the first place. These bits may actually need quite a bit of stack as well. Thus, I still think that we should try to reset to the top of the stack if possible.